src/plugins/analysis/plugin.py

```diff
@@ -0,0 +1,395 @@
+
+    def __init__(
+        self,
+        plugin: PluginV1,
+        config: Config,
+        in_queue: Queue[PluginRunner.Task],
+        out_queue: Queue[str],
+        stats: Array[ctypes.c_float],
+        stats_count: Value[int],
+        stats_idx: Value[int],
+    ):
+        # Queue[..] works: https://github.com/python/cpython/pull/19423
+        # pylint: disable=unsubscriptable-object
+        self._plugin = plugin
+        self._config = config
+
+        self._in_queue = in_queue
+        self._in_queue.close()
+        self._out_queue = out_queue
+
+        self._stats = stats
+        self._stats_count = stats_count
+        self._stats_idx = stats_idx
+
+        self._process = Process(
+            target=self._entrypoint,
+            name=f'{plugin.metadata.name} worker',
+        )
+
+    def start(self):
+        self._process.start()
+
+    def terminate(self):
+        self._process.terminate()
+
+    def _entrypoint(self):
+        """Entrypoint for the worker processes.
+        Pulls :py:class:`PluginRunner.Task` from ``_in_queue`` and writes finished analyses to ``_out_queue``.
+        """
+        # pylint: disable=too-complex
+        run = True
+
+        def _handle_sigterm(signum, frame):
+            del signum, frame
+            logging.critical(f'{self._process} received SIGTERM. Shutting down.')
+            nonlocal run
+            # Let the Worker finish the analysis it is currently doing
+            run = False
+
+        signal.signal(signal.SIGTERM, _handle_sigterm)
+
+        def _raise_timeout(signum, frame):
+            del signum, frame
+            raise Worker._TimeoutException()
+
+        signal.signal(signal.SIGALRM, _raise_timeout)
+
+        while run:
+            try:
+                # We must have some non infinite delay here to avoid blocking even after _handle_sigterm is called
+                task = self._in_queue.get(self._config.delay)
+            except queue.Empty:
+                continue
+
+            entry = {}
+            try:
+                logging.debug(f'{self._process}: Begin {self._plugin.metadata.name} analysis on {task.uid}')
+                start_time = time.time()
+
+                signal.alarm(self._config.timeout)
+                analysis = self._plugin.get_analysis(io.FileIO(task.path), task.virtual_file_path, task.dependencies)
+                signal.alarm(0)
+
+                duration = time.time() - start_time
+
+                # Why can't python have tagged unions...
+                entry['analysis'] = analysis
+                logging.debug(f'{self._process}: Finished {self._plugin.metadata.name} analysis on {task.uid}')
+                if duration > 120:
+                    logging.info(
+                        f'Analysis {self._plugin.metadata.name} on {task.uid} is slow: took {duration:.1f} seconds'
+                    )
+                self._update_duration_stats(duration)
+            except Worker._TimeoutException:
```

Comment:
If a timeout happens we have to kill all children of the worker to prevent zombies.